### PR TITLE
 InstructorHome: load courses incrementally by ajax #4108 

### LINKED ISSUE #4108: InstructorHome: load courses incrementally by ajax

### PR DESCRIPTION
**Pull request by yj-soh**
Fixes #4108 

@kanghj 
I'm not too sure about the *Logic classes, so could you take a closer look over those at least?

### CODE REVIEW

**Comment by damithc** [comment]
Do we load all automatically or just the most recent n courses?

**Comment by yj-soh** [comment]
It should currently automatically load only the top 3 (via `COURSE_PANELS_TO_AUTO_LOAD_COUNT ` in `instructorHome.js`) as per our discussion.
The specific courses depend on the sort order.
Hidden courses are indicated with a down arrow on the right.

**Comment by damithc** [comment]
Ok, that's good.

File: src/main/webapp/js/instructorHome.js [hunk 1]
```diff
@@ -1,27 +1,61 @@
...
+                type : 'POST',
+                url : '/page/instructorHomePage?' + formData,
+                beforeSend : function() {
+                    $(content).html("<img src='/images/ajax-loader.gif'/>");
+                },
+                error : function() {
+                    // TODO: show an error message?         
+                },
+                success : function(data) {
+                    // .replaceWith() fixes <br> in statuses' tooltips by closing the tooltip <span>
```
Comment (kanghj) [inline src/main/webapp/js/instructorHome.js]: Don't write comments for the reviewer. Future developers will not understand this comment

File: src/main/webapp/js/instructorHome.js [hunk 2]
```diff
@@ -1,27 +1,61 @@
+var COURSE_PANELS_TO_AUTO_LOAD_COUNT = 3;
+
 $(document).ready(function() {
     setupFsCopyModal();
     
     //Click event binding for radio buttons
-    var radiobuttons = $("label[name='sortby']");
-    $.each(radiobuttons, function() {
+    var radioButtons = $('label[name="sortby"]');
```
Comment (kanghj) [inline src/main/webapp/js/instructorHome.js]: Refrain from refactoring unrelated code. Anyway this should be prefixed with a $

File: src/main/webapp/js/instructorHome.js (hunk 1 above)
Comment (yj-soh) [inline src/main/webapp/js/instructorHome.js]: It's not meant for the reviewer.
It's meant to explain why it's using `$panel[0].outerHTML = data;` instead of a more conventional `$panel.replaceWith(data)` in case anyone plans on refactoring it in the future.
I suppose it could be made more clear, though.

File: src/main/java/teammates/ui/controller/InstructorHomeCourseAjaxPageData.java [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+                    className + "notify-pending-comments-for-test",
+                    getInstructorStudentCommentClearPendingLink(courseId),
+                    "Send email notification to recipients of " + pendingCommentsCount
+                        + " pending " + (pendingCommentsCount > 1 ? "comments" : "comment"));
+    
+            return Arrays.asList(enroll, view, edit,add, archive, pending, delete);
+        }
+    }
+    
+    private List<Map<String, Object>> createSessionRows(List<FeedbackSessionAttributes> sessions,
```
Comment (kanghj) [inline src/main/java/teammates/ui/controller/InstructorHomeCourseAjaxPageData.java]: This seems unrelated to your pr, but I personally find this <code>Map\<String, Object\></code> worrying.

File: src/main/webapp/js/instructorHome.js (hunk 1 above)
Comment (kanghj) [inline src/main/webapp/js/instructorHome.js]: Yes, make it clearer then

File: src/main/java/teammates/ui/controller/InstructorHomeCourseAjaxPageData.java (hunk 3 above)
Comment (yj-soh) [inline src/main/java/teammates/ui/controller/InstructorHomeCourseAjaxPageData.java]: Yes, I do want to refactor the instructor home files, they were the first ones done, so they didn't benefit from the experience from the rest of the code written during the JSTL migration.

Should I make an issue for it? "Refactor InstructorHome" or something?

Comment (damithc) [inline src/main/java/teammates/ui/controller/InstructorHomeCourseAjaxPageData.java]: Yes, let's do it as a separate PR.

**Comment by kanghj** [comment]
Behaviour is fine.

**Comment by yj-soh** [comment]
I forgot to ask this, but what should be the expected behaviour on failure?
If a person were to load the page, then disconnect before loading some course table, what should happen?

**Comment by damithc** [comment]
Show an error message '...Failed to load... retry' (similar to what we do when the session stats fail to load)? The question is where to show the message. May be in the panel heading?

**Comment by yj-soh** [comment]
A couple of sketch-ups.

---

`<strong>` tag in the panel header.

* Whole panel is getting replaced, so it'll disappear once it's loaded.

---

Utilizing the `statusMessage` div.

* Append whenever an error is detected?
* Hide upon clicking any unloaded panel?
* Or a line-by-line check for the specific course with an error to remove it from the message?)

**Comment by damithc** [comment]
I like the first option because the error appears in the context where it
happened. If only if we can make it a more noticeable. Any chance we can
make the letters yellow? Or make the background of the letters red? Or show
an error icon in red together with the white text?
Click to retry --> Click here to retry?

**Comment by yj-soh** [comment]
I personally believe just adding an icon in white is sufficient to get the point across.


A background colour just generally looks bad.

**Comment by damithc** [comment]
Yes, an icon is enough. Can you make the background of the icon white? Red
and blue is hard on the eye when they are touching each other.
Alternatively, try making the icon orange.

**Comment by yj-soh** [comment]
Making the background white would result in a box of white, not just the insides of the icon.
Orange is like so:

**Comment by damithc** [comment]
The icon is not very clear, but I guess our choices are limited. May be put
it inside the square brackets? We can also underline 'here' to make it look
clickable.
The orange used is the standard bootstrap 'warning' orange right?
Instead of the icon, we can try making the text orange, possibly including
the down arrow.
Anyway, do whatever you think looks better. :-)

**Comment by yj-soh** [comment]
@kanghj Another review, please. Not too sure about the test, so please take a look over that commit.
